osu.Game/Rulesets/Edit/HitObjectComposer.cs

```diff
@@ -388,8 +387,7 @@ public override SnapResult SnapScreenSpacePositionToValidTime(Vector2 screenSpac
     /// A non-generic definition of a HitObject composer class.
     /// Generally used to access certain methods without requiring a generic type for <see cref="HitObjectComposer{T}" />.
     /// </summary>
-    [Cached(typeof(HitObjectComposer))]
-    [Cached(typeof(IPositionSnapProvider))]
+    [Cached]
```

Comment:
Caching as most derived would mean children won't be able to DI the composer via `HitObjectComposer`, is this intended? at best, it should be:
```suggestion
    [Cached]
    [Cached(typeof(HitObjectComposer))]
```

Reply:
Honestly, if it doesn't fail in any existing cases, then I'd be in favour of not caching as `HitObjectComposer` and always caching as interface types. Helps test stuff and substituting own implementations if needed.

It doesn't seem to have bothered CI, but I'm not sure that we have 100% coverage there so I'll attempt to check manually...

Reply:
I've ~~wasted~~ spent some time looking at DI internals and this works as intended. Putting `[Cached]` at the HOC level is the same thing as putting `[Cached(typeof(HitObjectComposer))]` because dependency activators [are collected from all base classes in a tree-like manner](https://github.com/ppy/osu-framework/blob/d62a9a43827b5df3761cfee45d797decd0644bd0/osu.Framework/Allocation/DependencyActivator.cs#L45-L46):

![image](https://user-images.githubusercontent.com/20418176/166895591-0d4409fd-8638-45f6-8913-1b405ebf0f3b.png)

Also the composer caches one of these, too:

https://github.com/ppy/osu/blob/2822ae09811bd6d1b4f807afbdca7ad7d047b236/osu.Game/Screens/Edit/Compose/ComposeScreen.cs#L46-L48

I don't think we're fixing how DI works here so I'd say it's probably fine to proceed with this as is...?